File: python/ai_evidence/patterns.py
"""Extraction behaviour, loaded from the shared definition.

Both implementations read ``shared/patterns.json`` so a pattern can only be
changed in one place. The two regex engines spell Unicode classes differently,
so the shared file uses placeholders and each language substitutes its own
form: JavaScript uses ``\\p{Letter}`` with the ``u`` flag, Python uses
``[^\\W\\d_]``.
"""
from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Pattern

_HERE = Path(__file__).resolve().parent
_CANDIDATES = [
    _HERE / "patterns.json",                       # packaged copy
    _HERE.parent.parent / "shared" / "patterns.json",  # repository checkout
]


def _load() -> dict:
    for p in _CANDIDATES:
        if p.exists():
            return json.loads(p.read_text(encoding="utf-8"))
    raise FileNotFoundError("shared/patterns.json not found")


PATTERNS = _load()

_SUBSTITUTIONS = {
    "{{LETTER}}": r"[^\W\d_]",
    "{{NUMBER}}": r"\d",
    "{{UPPER}}": r"[A-ZÀ-ÞĀ-Ž]",
}

_FLAGS = {"i": re.IGNORECASE, "m": re.MULTILINE, "s": re.DOTALL, "u": re.UNICODE}


def rx(source: str, flags: str = "i") -> Pattern[str]:
    for token, replacement in _SUBSTITUTIONS.items():
        source = source.replace(token, replacement)
    bits = 0
    for f in flags:
        bits |= _FLAGS.get(f, 0)
    return re.compile(source, bits)


T = PATTERNS["thresholds"]
CLAIM_PATTERNS = [(t, rx(src)) for t, src in PATTERNS["claimPatterns"]]
GATES = {k: rx(v) for k, v in PATTERNS["gates"].items() if k != "$comment"}
SENTENCE = {k: rx(v["source"], v["flags"]) for k, v in PATTERNS["sentences"].items()}
HTML_SETS = {k: set(v) for k, v in PATTERNS["html"].items()}
ATOMIC_TYPES = set(PATTERNS["atomicTypes"])
SCHEMA_TYPE_MAP = PATTERNS["schemaTypeMap"]

File: reference-implementation/patterns.js
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

/**
 * Extraction behaviour is defined once, in shared/patterns.json, and loaded by
 * every implementation. Keeping it as data rather than code is what stops the
 * JavaScript and Python ports drifting: a pattern can only be changed in one
 * place, and the conformance corpus proves both engines still agree.
 *
 * The two regex engines spell Unicode classes differently, so the shared file
 * uses placeholders and each language substitutes its own form.
 */
const here = path.dirname(fileURLToPath(import.meta.url));
export const PATTERNS = JSON.parse(
  fs.readFileSync(path.join(here, '..', 'shared', 'patterns.json'), 'utf8')
);

const substitute = src => src
  .replaceAll('{{LETTER}}', '\\p{Letter}')
  .replaceAll('{{NUMBER}}', '\\p{Number}')
  .replaceAll('{{UPPER}}', '\\p{Uppercase_Letter}');

export const rx = (src, flags = 'i') => new RegExp(substitute(src), flags);

export const T = PATTERNS.thresholds;
export const CLAIM_PATTERNS = PATTERNS.claimPatterns.map(([type, src]) => [type, rx(src)]);
export const GATES = Object.fromEntries(
  Object.entries(PATTERNS.gates).filter(([k]) => k !== '$comment').map(([k, v]) => [k, rx(v)])
);
// Flags are declared alongside each pattern. Case-sensitivity is part of what
// a pattern means, not a detail of how it is compiled: defaulting `initial` to
// the i flag made it match the possessive in "the substrate's." and silently
// merged sentences.
export const SENTENCE = Object.fromEntries(
  Object.entries(PATTERNS.sentences).map(([k, v]) => [k, rx(v.source, v.flags)])
);
export const HTML_SETS = Object.fromEntries(
  Object.entries(PATTERNS.html).map(([k, v]) => [k, new Set(v)])
);
export const ATOMIC_TYPES = new Set(PATTERNS.atomicTypes);
export const SCHEMA_TYPE_MAP = PATTERNS.schemaTypeMap;
